### PR TITLE
introduce v1alpha2 api with auto conversion

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

This is related to #1526  and #1379.

This introduces a new api version `v1alpha2` while still allowing user
to use `v1alpha1` API.

- Only one version of the CRD can be stored, so `v1alpha1` is the one
stored
- All the version of the CRD **must** be schema compatible (before 1.16)
  - `v1alpha1` gets **all** the fields, meaning :
    - the new ones (from `v1alpha2`, …)
    - the old deprecated ones (that will be removed from `v1alpha2`)
  - we may *only* document the "frozen" `v1alpha1` field (and not the
    new one that are coming from the other api versions)
- `kubectl` get returns the highest priority version, aka the most
  recent one, here `v1alpha2`. To make *sure* any object created from
  any version are valid (while getting), we convert them using the
  admission webhook (#blackmagic 🧙).
  - Object created using `v1alpha1` API will automatically be
  converted (see below)
  - This has the downsize of not being able to `get` v1alpha1 object
  anymore (although it should be possible using the `client-go` and
  some even more #blackmagic 🧙 hack on the webhook side probably)
- One benefit of this approach (`apis.Convertible`) is that once
  we *only* support 1.16 (the release with conversion webhook), this
  will be usable almost as is (for the conversion webhook) 👼.

This is done in a quick and dirty way:
- Lot's of code is duplicated where it would not be needed, it was
  just quicker to do so at that point in time
- I completely skipped the *variable interpolation* validation for
  `Parameters` as those values would/will need to be automatically
  transformed too (but it was a bit complicated for what I wanted to
  experiment on)
- I did not implement the **reconcilier** part of this. It should be
  simple as:
  - it will use only one object (the store one, aka `v1alpha1`)
  - it will make the assumption that the objects are fully
    transformed (either by the webhook or in-memory by the
    controller), meaning there will be no *handling
    backward-compatible* piece of code in the reconcilier (all is
    contained in the `Validable`, `Defaultable` and `Convertible` impl.).
- I did not run/fix the tests :wink: 

Possible next steps are:
- Split this into smaller PR.
- Introduce a `v1alpha2` package early but not exposing it before we
  are ready to. This will allow us to not have to rush this for
  0.9, **and** to increment over this in smaller step.
- Decide what `v1alpha2` should look like — I did this experiment with
  `TaskSpec.Inputs.Params` to `TaskSpec.Params` but we should decide
  what we want to change and target that.
- Reduce duplication by reusing anything that didn't change from
  `v1alpha1` to `v1alpha2` in `v1alpha1` (aka refering the same type
  in go)
- Related to this experiment, implement the variable interpolation
  validation thing **and** the reconcilier part too.
- Need to figure out if the definition types (`Task`, `Pipeline`, …)
  need a `controller` (I don't think so but…)

Example of conversion
```yaml
#This task deploys with kubectl apply -f <filename>
apiVersion: tekton.dev/v1alpha1
kind: Task
metadata:
  name: demo-deploy-kubectl
spec:
  inputs:
    resources:
    - name: workspace
      type: git
    - name: image
      type: image
    params:
    - name: path
      description: Path to the manifest to apply
    - name: yqArg
      description: Okay this is a hack, but I didn't feel right hard-codeing `-d1` down below
    - name: yamlPathToImage
      description: The path to the image to replace in the yaml manifest (arg to yq)
  steps:
  - name: replace-image
    image: mikefarah/yq
    command: ['yq']
    args:
    - "w"
    - "-i"
    - "$(inputs.params.yqArg)"
    - "$kube(inputs.params.path)"
    - "$(inputs.params.yamlPathToImage)"
    - "$(inputs.resources.image.url)"
  - name: run-kubectl
    image: lachlanevenson/k8s-kubectl
    command: ['kubectl']
    args:
    - 'apply'
    - '-f'
    - '$(inputs.params.path)'
```

becomes the following at creation
```yaml
#This task deploys with kubectl apply -f <filename>
apiVersion: tekton.dev/v1alpha2
kind: Task
metadata:
  name: demo-deploy-kubectl
spec:
  inputs:
    resources:
    - name: workspace
      type: git
    - name: image
      type: image
  params:
  - name: path
    description: Path to the manifest to apply
  - name: yqArg
    description: Okay this is a hack, but I didn't feel right hard-codeing `-d1` down below
  - name: yamlPathToImage
    description: The path to the image to replace in the yaml manifest (arg to yq)
  steps:
  - name: replace-image
    image: mikefarah/yq
    command: ['yq']
    args:
    - "w"
    - "-i"
    - "$(inputs.params.yqArg)"
    - "$kube(inputs.params.path)"
    - "$(inputs.params.yamlPathToImage)"
    - "$(inputs.resources.image.url)"
  - name: run-kubectl
    image: lachlanevenson/k8s-kubectl
    command: ['kubectl']
    args:
    - 'apply'
    - '-f'
    - '$(inputs.params.path)'
```

**Putting an hold as this is not meant to be merged, ever !**

/hold
/cc @bobcatfish @skaegi @ImJasonH @afrittoli @abayer 

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

### CODE REVIEW

File: cmd/webhook/main.go [hunk 1]
```diff
@@ -104,7 +113,7 @@ func main() {
 
 	// Decorate contexts with the current state of the config.
 	ctxFunc := func(ctx context.Context) context.Context {
-		return v1alpha1.WithDefaultConfigurationName(store.ToContext(ctx))
+		return v1alpha2.WithUpgradeViaDefaulting(store.ToContext(ctx))
```
Comment (vdemeester) [inline cmd/webhook/main.go]: This tells the webhook to do the conversion :angel:

File: config/300-clustertask.yaml [hunk 2]
```diff
@@ -17,6 +17,13 @@ metadata:
   name: clustertasks.tekton.dev
 spec:
   group: tekton.dev
+  versions:
+  - name: v1alpha1
+    served: true
+    storage: true
+  - name: v1alpha2
+    served: true
+    storage: false
```
Comment (vdemeester) [inline config/300-clustertask.yaml]: `v1alpha1` is the stored version, but both are served

File: examples/pipelineruns/pipelinerun.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 ---
```
Comment (vdemeester) [inline examples/pipelineruns/pipelinerun.yaml]: This files contains `v1alpha2` and `v1alpha1` Tasks

File: hack/update-codegen.sh [hunk 4]
```diff
@@ -42,7 +42,7 @@ ${GOPATH}/bin/deepcopy-gen \
 # Knative Injection
 ${KNATIVE_CODEGEN_PKG}/hack/generate-knative.sh "injection" \
   github.com/tektoncd/pipeline/pkg/client github.com/tektoncd/pipeline/pkg/apis \
-  "pipeline:v1alpha1" \
+  "pipeline:v1alpha1,v1alpha2" \
```
Comment (vdemeester) [inline hack/update-codegen.sh]: This updates codegen to generate v1alpha1 and v1alpha2 clientset, informers, …

File: pkg/apis/pipeline/v1alpha1/task_conversion.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha2"
+	"knative.dev/pkg/apis"
+)
+
+func (source *Task) ConvertUp(ctx context.Context, obj apis.Convertible) error {
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/task_conversion.go]: This allows us to take a `v1alpha1` object and "convert it up" to a more recent version (that will be passed, by the webhook or the reconcilier for in-memory upgrade)

File: pkg/apis/pipeline/v1alpha1/task_conversion.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	case *v1alpha2.Task:
+		sink.ObjectMeta = source.ObjectMeta
+		return source.Spec.ConvertUp(ctx, &sink.Spec)
+	default:
+		return fmt.Errorf("unknown version, got: %T", sink)
+	}
+}
+
+func (sink *Task) ConvertDown(ctx context.Context, obj apis.Convertible) error {
+	switch source := obj.(type) {
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/task_conversion.go]: This does the opposite of `ConvertUp`, it takes a "higher" version of the Task object and converts it "back" to `v1alpha1` (hence the need for `v1alpha1` types to have *all* the fields

File: pkg/apis/pipeline/v1alpha1/task_defaults.go [hunk 7]
```diff
@@ -26,13 +29,27 @@ func (t *Task) SetDefaults(ctx context.Context) {
...
+			fmt.Printf("v: %+v\n", v)
+			fmt.Printf("v.inputs: %+v\n", v.Inputs)
+			fmt.Printf("v.outputs: %+v\n", v.Outputs)
+			alpha1 := TaskSpec{}
+			if alpha1.ConvertDown(ctx, v) == nil {
+				*ts = alpha1
+			}
+		}
+	}
+	fmt.Printf("ts: %+v\n", ts)
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/task_defaults.go]: This combined with https://github.com/tektoncd/pipeline/pull/1529/files#diff-58de452513b7e8d8d3cfea23eb4ae6a8R116 make the conversion automatic from `v1alpha1` to a `v1alpha2` compatible object (still stored as `v1alpha1`)

File: pkg/apis/pipeline/v1alpha1/task_types.go [hunk 8]
```diff
@@ -44,6 +44,11 @@ type TaskSpec struct {
 	// Task is run.
 	// +optional
 	Outputs *Outputs `json:"outputs,omitempty"`
+	// Params is a list of input parameters required to run the task. Params
+	// must be supplied as inputs in TaskRuns unless they declare a default
+	// value.
+	// +optional
+	Params []ParamSpec `json:"params,omitempty"`
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/task_types.go]: This fields comes from `v1alpha2` "struct" and "might not be documented"

File: pkg/apis/pipeline/v1alpha1/task_types.go [hunk 9]
```diff
@@ -104,11 +110,13 @@ type Inputs struct {
...
 	Resources []TaskResource `json:"resources,omitempty"`
-	// Params is a list of input parameters required to run the task. Params
+
+	// Depraceted
+	// DeprecatedParams is a list of input parameters required to run the task. Params
 	// must be supplied as inputs in TaskRuns unless they declare a default
 	// value.
 	// +optional
-	Params []ParamSpec `json:"params,omitempty"`
+	DeprecatedParams []ParamSpec `json:"params,omitempty"`
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/task_types.go]: Renaming it makes it easier to know it's deprecated and we shouldn't depend/looking at it in the reconcilier

File: pkg/apis/pipeline/v1alpha1/task_validation.go [hunk 10]
```diff
@@ -95,9 +95,11 @@ func (ts *TaskSpec) Validate(ctx context.Context) *apis.FieldError {
...
 	}
 
-	if err := validateInputParameterVariables(ts.Steps, ts.Inputs); err != nil {
-		return err
-	}
+	/*
+		 	if err := validateInputParameterVariables(ts.Steps, ts.Inputs); err != nil {
+				return err
+			}
+	*/
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/task_validation.go]: As said in the comments, this is commented to skip the "auto" migrate variable interpolation

**Comment by vdemeester** [comment]
btw, this is how `knative/serving` does it :angel: :wink:

**Comment by vdemeester** [comment]
> * Split this into smaller PR.

I'll start tackling this next week, especially the following items :angel: 

> * Introduce a `v1alpha2` package early but not exposing it before we
>   are ready to. This will allow us to not have to rush this for
>   0.9, **and** to increment over this in smaller step.
> * Reduce duplication by reusing anything that didn't change from
>   `v1alpha1` to `v1alpha2` in `v1alpha1` (aka refering the same type
>   in go)

File: examples/pipelineruns/pipelinerun.yaml [hunk 11]
```diff
@@ -131,7 +131,7 @@ spec:
     - "w"
     - "-i"
     - "$(inputs.params.yqArg)"
-    - "$(inputs.params.path)"
+    - "$kube(inputs.params.path)"
```
Comment (cccfeng) [inline examples/pipelineruns/pipelinerun.yaml]: Hi, it looks like this line of code has one more word “kube” ?

**Comment by vdemeester** [comment]
I am going to close as split work has started :wink: 
/close

**Pull request closed without merging**